morpho-checkpoint-node/tests/Test/Morpho/QSM.hs

```diff
@@ -335,46 +354,98 @@ prop_4 = noShrinking $
           majority = 2
         }
 
-setup :: Int -> Int -> IO [NodeHandle]
-setup nodesNum testId = do
+setup :: Int -> Int -> Int -> IO [NodeHandle]
+setup nodesNum requiredMajority testId = do
   let testDir = mkTestDir testId
   removePathForcibly testDir
   createDirectoryIfMissing True testDir
-  forM [0 .. (nodesNum - 1)] $ runDualNode True testId
+  forM [0 .. (nodesNum - 1)] $ runDualNode True nodesNum requiredMajority testId
 
 cleanup :: NodeHandle -> IO ()
 cleanup NodeHandle {..} = do
   cancel mainNode
 
-runDualNode :: Bool -> Int -> Int -> IO NodeHandle
-runDualNode createDir testId nodeId = do
+runDualNode :: Bool -> Int -> Int -> Int -> Int -> IO NodeHandle
+runDualNode createDir nodesNum requiredMajority testId nodeId = do
```

Comment:
And this should probably be `Env -> IO NodeHandle` instead. You can introduce a function like `configToEnvs :: Config -> IO (Int -> Env)` to be able to call it from `setup` with a specific node index.

Note that I think it should be `IO (Int -> Env)` so that you can do a single initialization of especially the starting time, which should be the same across all nodes.